packages/react-compose/src/createOptionsResolver.ts

```diff
@@ -0,0 +1,65 @@
+import { ComposePreparedOptions, GenericDictionary } from './types';
+import { getNativeElementProps, assign } from '@uifabric/utilities';
+import { resolveClasses } from './resolveClasses';
+import { resolveSlotProps } from './resolveSlotProps';
+
+export type OptionsResolverResult = {
+  state: GenericDictionary;
+  slots: GenericDictionary;
+  slotProps: GenericDictionary;
+};
+
+/**
+ * A set of mapped props for intrinsic element types.
+ */
+export const defaultMappedProps: Record<string, string> = {
+  iframe: 'src',
+  img: 'src',
+  input: 'type',
+};
+
+/**
+ * Creates an options resolve function which should attach to an options object
+ * for a composed component.
+ */
+export const createOptionsResolver = <TState>(options: ComposePreparedOptions) => {
+  // Returning a function so that "resolve" function shows up in profiling.
+  // tslint:disable-next-line:no-function-expression
+  return function resolve(stateOrProps: TState): OptionsResolverResult {
+    const slotProps: OptionsResolverResult['slotProps'] = {};
+    const state = stateOrProps as OptionsResolverResult['state'];
+    const slots: OptionsResolverResult['slots'] = {
+      ...options.slots,
+      root: state.as || options.slots.root || 'div',
+    };
+    const result: OptionsResolverResult = {
+      state,
+      slotProps,
+      slots,
+    };
+
+    // Mix unrecognized props onto root, excluding the handled props.
+    assignToMapObject(
+      slotProps,
+      'root',
+      getNativeElementProps(state.as, state, [...options.handledProps, 'className']),
+    );
+
+    // Resolve slotProps/slots from state.
+    resolveSlotProps(result);
```

Comment:
It looks like resolveSlotProps does return result, but it doesn't need to because it mutates the result. Instead, you could pass state, slotProps, and slots down and return a result and have that get set here.

Reply:
I would normally do that, but I've optimized to be able to chain resolvers to mutate an ongoing result for performance reasons. Doing otherwise would mean each resolver would need to allocate a new object and mix the old result. There isn't really a need to do so in an internal function.